jetty-core/jetty-io/src/main/java/org/eclipse/jetty/io/content/ContentSourcePublisher.java

```diff
@@ -13,147 +13,311 @@
-        private long demand;
-        private boolean stalled;
-        private boolean cancelled;
-        private boolean terminated;
+        // As per rule 1.9, we need to throw a `java.lang.NullPointerException`
+        // if the `Subscriber` is `null`
+        if (subscriber == null)
+        {
+            NullPointerException error = new NullPointerException("Flow.Subscriber must not be null");
+            content.fail(error);
+            throw error;
+        }
+        LastWillSubscription subscription = new ActiveSubscription(content, subscriber);
+        // As per rule 1.9, this method must return normally (i.e. not throw).
+        try
+        {
+            subscriber.onSubscribe(subscription);
+        }
+        catch (Throwable err)
+        {
+            // As per rule 2.13, we MUST consider subscription cancelled and
+            // MUST raise this error condition in a fashion that is adequate for the runtime environment.
+            subscription.cancel(err, LastWillSubscription.FinalSignal.SUPPRESS);
+            LOG.error("Flow.Subscriber " + subscriber + " violated rule 2.13", err);
+        }
+    }
 
-        public SubscriptionImpl(Content.Source content, Flow.Subscriber<? super Content.Chunk> subscriber)
+    private void onMultiSubscribe(Flow.Subscriber<? super Content.Chunk> subscriber)
+    {
+        // As per rule 1.9, we need to throw a `java.lang.NullPointerException`
+        // if the `Subscriber` is `null`
+        if (subscriber == null)
         {
-            this.content = content;
-            this.subscriber = subscriber;
-            this.stalled = true;
+            throw new NullPointerException("Flow.Subscriber must not be null");
         }
+        LastWillSubscription subscription = new ExhaustedSubscription();
 
+        // As per 1.9, this method must return normally (i.e. not throw).
+        try
+        {
+            // As per rule 1.9, the only legal way to signal about Subscriber rejection
+            // is by calling onError (after calling onSubscribe).
+            subscriber.onSubscribe(subscription);
+            subscriber.onError(new IllegalStateException("Content.Source was exhausted."));
+        }
+        catch (Throwable err)
+        {
+            // As per rule 2.13, we MUST consider subscription cancelled and
+            // MUST raise this error condition in a fashion that is adequate for the runtime environment.
+            subscription.cancel(err, LastWillSubscription.FinalSignal.SUPPRESS);
+            LOG.error("Flow.Subscriber " + subscriber + " violated rule 2.13", err);
+        }
+    }
+
+    // Subscription that have a reason of the cancellation and information about
+    // final signal (last will and testament pattern) that must be reported to the Subscriber.
+    // Relates to https://github.com/reactive-streams/reactive-streams-jvm/issues/271
+    private interface LastWillSubscription extends Flow.Subscription
```

Comment:
as much as I appreciate novelty names like "LastWill" (and have created a few in my time (e.g. EatWhatYouKill)), we do need to avoid names that might be triggering or off-topic in an open source project  like Jetty.   I think something like `AbstractSubscription` is more appropriate.